### ISSUE TITLE
To do list: Will

### ISSUE THREAD

**Issue by Outfenneced** [jnick1/mesa#58]
- [ ] Force python to run in background, allowing user to do other things during calculation. This will make it so extremely efficient calculation times aren't as necessary
- [ ] Reduce calls to attendee_available (Longest execution time, most calls)
- [ ] Improve calculation times for tracker_available_attendees

**Comment by jnick1** [comment]
I'm wondering something for tracker_check_required_busy. As per my current understanding, it checks if the required attendees are busy at an event time for a duration by essentially remembering which times it has checked before, which are all guaranteed to have greater durations (due to how we only decrease duration). Since it knows which times have been checked before, and since it knows that those are ok, it automatically knows that all shorter events at that time are also ok. Does that sum it up fairly well?

If this is correct, I'm wondering if you can implement the following idea. Currently, this tracker (while a fantastic idea) could be improved to catch even more redundant cases. Right now it only remembers that it has checked a time, but imagine this: if it checks noon for 1 hour, and that's good, then we already know that noon for 45, 30 and 15 minutes work (currently the tracker handles this). BUT, we also know that 12:15 for 45, 30, 15 minutes, and 12:30 for 30, 15 minutes and 12:45 for 15 minutes are also good. If implemented correctly, this could cut off massive chunks of time, as we could just skip (so to speak) to the point generation.

Some ideas for implementation: you could store the duration along with the time in the tracker, and then check "back" in the tracker for events with durations (up to the max, preferred, duration) that circumscribe whatever event time and duration you're currently checking. However, there may be more efficient methods for implementation, so you'll probably want to think up some on your own.

**Comment by Outfenneced** [comment]
Here's how tracker_check_required_busy works: Basically, the duration starts at full then decreases by an increment of granularity each time through the loop. The full duration goes through normally, adding available times for all required attendees to the list. The next duration down comes along, once it finds that the eventTime is in the tracker (therefore from the duration one increment of granularity larger), it finds the next event time by adding the granularity (line 92-93 of pointListGenerator) and also adds that to the tracker (that covers the 45, 30, 15 filling). If the eventTime is not in the tracker or the value of the element at the eventTime is false, that means some required attendee was unable to attend at the larger duration, so it recalculates for the lower duration that the loop is currently in.

The current implementation of tracker_check_required_busy, I believe, already fulfills the idea you have requested.

Interestingly, attempting to implement the same method with tracker_available_attendees somehow doubles calculation time.

**Comment by Outfenneced** [comment]
If my explanation of why it fills the hour isn't in-depth enough, think of it this way: the hour duration goes through normally on one specific hour. The 45 minute duration at the same time as the hour duration also forces the 45 minute duration an increment of 15 minutes forward in time. Those are the two 45 minute durations that fit. The second 45 minute duration indicator (during the 30 minute duration step) causes the tracker to have the time a half hour after the hour. This 30 minute one causes the 15 minute one (at the 45 minutes after the hour) to also be tracked as available. Because of the cascading downwards of these checks and the order in which times are checked, this causes all 4 15 minute apart times to be tracked before are checked.

**Comment by Outfenneced** [comment]
Actually, huh, looking back through tracker_check_required_busy, I think I may have mistyped something that may have caused the decreased calculation times. I tend to accidentally put a return of "true" as required attendees being available. I'm rerunning tests to make sure my outputs from the function are good.

**Comment by jnick1** [comment]
I'm working on a big word document now that runs through all the issues I'm looking at in the function. Just hold tight for that, and I'll upload it here soon.

**Comment by Outfenneced** [comment]
Will do!

**Comment by Outfenneced** [comment]
Yeah, looking through my testing outputs it looks like I definitely did something wrong. I'll wait for your post to make sure we're seeing the same thing. Definitely I made an error.

**Comment by jnick1** [comment]
Here's what I've found, it's 3 pages of dense stuff, but I think (and hope) it's all correct. It's a walkthough of the code and it's effects/results. I haven't included any suggestions to fix the issues found, but I'm sure you'll be able to think them up (plus I just wanted to be done with that document, jeez that took a while!). Forgive any spelling/grammar errors, I turned off the checks for those so it wouldn't spam my code samples with errors.

[tracker_check_required_busy comments.docx](https://github.com/jnick1/mesa/files/243431/tracker_check_required_busy.comments.docx)

**Comment by Outfenneced** [comment]
Currently performing a personal refactor of the classes file to remove some major issues I think I'm seeing in there. It'll be on a fork on my own repository so as to not effect this in any way. If I find it's better, I'll pull request it and have you ( @jnick1 ) look at it.
